File: test/git.test.ts
import { spawnSync } from "node:child_process";
import { mkdirSync, mkdtempSync, realpathSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { afterEach, describe, expect, it, vi } from "vitest";
import { commit, excludedFiles, repoRoot, stagedDiff, stagedStat, stagedStats } from "../src/git";

function git(cwd: string, ...args: string[]): string {
  const result = spawnSync("git", args, { cwd, encoding: "utf8" });
  if (result.status !== 0) {
    throw new Error(`git ${args.join(" ")}: ${result.stderr}`);
  }
  return result.stdout;
}

function newRepo(): string {
  const dir = realpathSync(mkdtempSync(path.join(tmpdir(), "commitron-git-")));
  git(dir, "init", "--quiet");
  git(dir, "config", "user.email", "test@example.com");
  git(dir, "config", "user.name", "Test");
  git(dir, "config", "commit.gpgsign", "false");
  return repoRoot(dir);
}

function stage(dir: string, name: string, content: string): void {
  const file = path.join(dir, name);
  mkdirSync(path.dirname(file), { recursive: true });
  writeFileSync(file, content);
  git(dir, "add", name);
}

afterEach(() => {
  vi.unstubAllEnvs();
});

describe("repoRoot", () => {
  it("fails outside a repository", () => {
    const dir = realpathSync(mkdtempSync(path.join(tmpdir(), "commitron-norepo-")));
    vi.stubEnv("GIT_CEILING_DIRECTORIES", path.dirname(dir));
    expect(() => repoRoot(dir)).toThrow();
  });
});

describe("stagedStats", () => {
  it("counts files and lines", () => {
    const dir = newRepo();
    expect(stagedStats(dir)).toEqual({ files: 0, added: 0, removed: 0 });

    stage(dir, "a.txt", "one\ntwo\nthree\n");
    stage(dir, "b.txt", "four\n");
    expect(stagedStats(dir)).toEqual({ files: 2, added: 4, removed: 0 });
  });
});

describe("stagedDiff", () => {
  it("honours exclusions", () => {
    const dir = newRepo();
    stage(dir, "src.go", "package main\n");
    stage(dir, "pnpm-lock.yaml", "lockfileVersion: 9\nnoise: everywhere\n");

    expect(stagedDiff([], dir)).toContain("lockfileVersion");

    const filtered = stagedDiff(["pnpm-lock.yaml"], dir);
    expect(filtered).not.toContain("lockfileVersion");
    expect(filtered).toContain("package main");
  });

  it("matches globs at any depth", () => {
    const dir = newRepo();
    stage(dir, "src.go", "package main\n");
    stage(dir, "nested/deep/thing.lock", "generated\n");

    expect(stagedDiff(["*.lock"], dir)).not.toContain("generated");
  });
});

describe("excludedFiles", () => {
  it("lists what the patterns dropped", () => {
    const dir = newRepo();
    stage(dir, "src.go", "package main\n");
    stage(dir, "pnpm-lock.yaml", "lockfileVersion: 9\n");

    expect(excludedFiles(["pnpm-lock.yaml"], dir)).toEqual(["pnpm-lock.yaml"]);
    expect(excludedFiles([], dir)).toEqual([]);
  });
});

describe("stagedStat", () => {
  it("keeps excluded files in the list", () => {
    const dir = newRepo();
    stage(dir, "src.go", "package main\n");
    stage(dir, "pnpm-lock.yaml", "lockfileVersion: 9\n");

    expect(stagedStat(dir)).toContain("pnpm-lock.yaml");
  });
});

describe("commit", () => {
  it("writes the message and empties the index", () => {
    const dir = newRepo();
    stage(dir, "a.txt", "hello\n");

    const sha = commit(
      "feat: add a.txt\n\n- because tests need a subject",
      { edit: false, verify: true },
      dir,
    );
    expect(sha.length).toBeGreaterThanOrEqual(6);

    const body = git(dir, "log", "-1", "--pretty=%B");
    expect(body).toContain("feat: add a.txt");
    expect(body).toContain("because tests need a subject");
    expect(stagedStats(dir).files).toBe(0);
  });
});

File: internal/gitx/git_test.go
package gitx

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

// newRepo creates a throwaway repository and makes it the working directory for
// the duration of the test. gitx talks to whatever repo the process is in.
func newRepo(t *testing.T) string {
	t.Helper()
	if _, err := exec.LookPath("git"); err != nil {
		t.Skip("git is not installed")
	}

	dir := t.TempDir()
	for _, args := range [][]string{
		{"init", "--quiet"},
		{"config", "user.email", "test@example.com"},
		{"config", "user.name", "Test"},
		{"config", "commit.gpgsign", "false"},
	} {
		cmd := exec.Command("git", args...)
		cmd.Dir = dir
		if out, err := cmd.CombinedOutput(); err != nil {
			t.Fatalf("git %s: %v\n%s", strings.Join(args, " "), err, out)
		}
	}

	previous, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	if err := os.Chdir(dir); err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { os.Chdir(previous) })

	// TempDir can be a symlinked path (/var vs /private/var on macOS); ask git.
	root, err := RepoRoot()
	if err != nil {
		t.Fatal(err)
	}
	return root
}

func stage(t *testing.T, name, content string) {
	t.Helper()
	path := filepath.Join(".", name)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatal(err)
	}
	cmd := exec.Command("git", "add", name)
	if out, err := cmd.CombinedOutput(); err != nil {
		t.Fatalf("git add %s: %v\n%s", name, err, out)
	}
}

func TestRepoRootFailsOutsideARepo(t *testing.T) {
	previous, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	dir := t.TempDir()
	// TMPDIR can itself live inside a repository (a GOTMPDIR pointed at the
	// working copy, for one). The ceiling stops git from walking up past it.
	t.Setenv("GIT_CEILING_DIRECTORIES", filepath.Dir(dir))
	if err := os.Chdir(dir); err != nil {
		t.Fatal(err)
	}
	defer os.Chdir(previous)

	if root, err := RepoRoot(); err == nil {
		t.Fatalf("expected an error outside a repository, got root=%s", root)
	}
}

func TestStagedStats(t *testing.T) {
	newRepo(t)

	got, err := StagedStats()
	if err != nil {
		t.Fatal(err)
	}
	if got.Files != 0 {
		t.Fatalf("an empty index should report 0 files, got %+v", got)
	}

	stage(t, "a.txt", "one\ntwo\nthree\n")
	stage(t, "b.txt", "four\n")

	got, err = StagedStats()
	if err != nil {
		t.Fatal(err)
	}
	want := Stats{Files: 2, Added: 4, Removed: 0}
	if got != want {
		t.Errorf("StagedStats() = %+v, want %+v", got, want)
	}
}

func TestStagedDiffHonoursExclusions(t *testing.T) {
	newRepo(t)
	stage(t, "src.go", "package main\n")
	stage(t, "pnpm-lock.yaml", "lockfileVersion: 9\nnoise: everywhere\n")

	full, err := StagedDiff(nil)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(full, "lockfileVersion") {
		t.Fatal("the unfiltered diff should contain the lockfile")
	}

	filtered, err := StagedDiff([]string{"pnpm-lock.yaml"})
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(filtered, "lockfileVersion") {
		t.Error("the lockfile should have been excluded")
	}
	if !strings.Contains(filtered, "package main") {
		t.Error("the real change should survive the exclusion")
	}
}

func TestStagedDiffGlobExcludesAtAnyDepth(t *testing.T) {
	newRepo(t)
	stage(t, "src.go", "package main\n")
	stage(t, "nested/deep/thing.lock", "generated\n")

	filtered, err := StagedDiff([]string{"*.lock"})
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(filtered, "generated") {
		t.Error("*.lock should match at any depth")
	}
}

func TestExcludedFiles(t *testing.T) {
	newRepo(t)
	stage(t, "src.go", "package main\n")
	stage(t, "pnpm-lock.yaml", "lockfileVersion: 9\n")

	got, err := ExcludedFiles([]string{"pnpm-lock.yaml"})
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 1 || got[0] != "pnpm-lock.yaml" {
		t.Errorf("ExcludedFiles() = %v, want [pnpm-lock.yaml]", got)
	}

	none, err := ExcludedFiles(nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(none) != 0 {
		t.Errorf("no patterns means nothing excluded, got %v", none)
	}
}

func TestStagedStatKeepsExcludedFiles(t *testing.T) {
	newRepo(t)
	stage(t, "src.go", "package main\n")
	stage(t, "pnpm-lock.yaml", "lockfileVersion: 9\n")

	stat, err := StagedStat()
	if err != nil {
		t.Fatal(err)
	}
	// The model should know the lockfile changed even though its diff is not sent.
	if !strings.Contains(stat, "pnpm-lock.yaml") {
		t.Error("the file list must still mention excluded files")
	}
}

func TestCommit(t *testing.T) {
	newRepo(t)
	stage(t, "a.txt", "hello\n")

	sha, err := Commit("feat: add a.txt\n\n- because tests need a subject", CommitOptions{Verify: true})
	if err != nil {
		t.Fatal(err)
	}
	if len(sha) < 6 {
		t.Errorf("Commit() returned %q, want a short sha", sha)
	}

	out, err := exec.Command("git", "log", "-1", "--pretty=%B").Output()
	if err != nil {
		t.Fatal(err)
	}
	body := string(out)
	if !strings.Contains(body, "feat: add a.txt") || !strings.Contains(body, "because tests need a subject") {
		t.Errorf("the message did not survive the commit:\n%s", body)
	}

	after, err := StagedStats()
	if err != nil {
		t.Fatal(err)
	}
	if after.Files != 0 {
		t.Errorf("the index should be empty after committing, got %+v", after)
	}
}
